compiler/ras/ILValidator.cpp

```diff
@@ -19,335 +19,243 @@
+   OMR::preCodegenValidationStrategy,
+   OMR::postILgenValidatonStrategy
+   };
 
-   ancestorNodes.remove(node);
+/**************************************************************************
+ *
+ * Implementation of TR::ILValidator
+ *
+ **************************************************************************/
+template <typename T, size_t N> static
+T* begin(T(&reqArray)[N])
+  {
+  return &reqArray[0];
+  }
+template <typename T, size_t N> static
+T* end(T(&reqArray)[N])
+   {
+   return &reqArray[0]+N;
    }
 
-void TR::ILValidator::soundnessRule(TR::TreeTop *location, bool condition, const char *formatStr, ...)
+TR::ILValidator::ILValidator(TR::Compilation *comp)
+   :_comp(comp)
    {
-   if (!condition)
-      {
-      if (location && location->getNode())
-         printDiagnostic("*** VALIDATION ERROR: IL is unsound at n%dn ***\nMethod: %s\n", location->getNode()->getGlobalIndex(), comp()->signature());
-      else
-         printDiagnostic("*** VALIDATION ERROR: IL is unsound ***\nMethod: %s\n", comp()->signature());
-      va_list args;
-      va_start(args, formatStr);
-      vprintDiagnostic(formatStr, args);
-      va_end(args);
-      printDiagnostic("\n");
-      FAIL();
-      }
+     /**
+      * All available `*ValidationRules` are initialized during the creation of
+      * ILValidator and they have the same lifetime as the ILValidator itself.
+      * It is during the call to `validate` where we decide which subset
+      * of the available ones we should use for Validation.
+      * This also removes the need to initialize a particular set of `Rule` objects
+      * every time a new Strategy is created, or a call to `validate` is made.
+      */
+     TR::MethodValidationRule* temp_method_rules[] =
+        {
+          new  (comp->trHeapMemory()) TR::SoundnessRule(_comp),
+          new  (comp->trHeapMemory()) TR::ValidateLivenessBoundaries(_comp)
+        };
+
+     TR::BlockValidationRule* temp_block_rules[] =
+        { new  (comp->trHeapMemory()) TR::ValidateNodeRefCountWithinBlock(_comp) };
+
+     TR::NodeValidationRule* temp_node_rules[] =
+        { 
+          new  (comp->trHeapMemory()) TR::ValidateChildCount(_comp),
+          new  (comp->trHeapMemory()) TR::ValidateChildTypes(_comp),
+          new  (comp->trHeapMemory()) TR::Validate_ireturnReturnType(_comp)
+        };
+     /**
+      * NOTE: Please initialize any new *ValidationRule here!
+      *
+      * Also, ILValidationRules.hpp and ILValidationStrategies.hpp
+      * need to be updated everytime a new ILValidation Rule
+      * is added.
+      */
+
+     _methodValidationRules.assign(begin(temp_method_rules), end(temp_method_rules));
+     _blockValidationRules.assign(begin(temp_block_rules), end(temp_block_rules));
+     _nodeValidationRules.assign(begin(temp_node_rules), end(temp_node_rules));
    }
 
-void TR::ILValidator::validityRule(Location &location, bool condition, const char *formatStr, ...)
+
+TR::Compilation *TR::ILValidator::comp()
```

Comment:
Shouldn't this be inline in the header file?

Reply:
Done.